Feat/remote notifications

DanGould opened the pull request.

DanGould committed:
add queued to v4

DanGould committed:
experiment with shorter timeouts

DanGould committed:
25 second Coinjoin on 7,7,7 CcjConfig

DanGould committed:
define AppleDeviceToken model

DanGould committed:
apply InitialCreate migration

DanGould committed:
control token db

DanGould commented:
@Kukks it seems like I'm connecting to a local db, but I create the migration manually with two `dotnet ef` commands to create the migration and update the db. I don't know if that's something that can be automated to just happen if it hasn't yet. Should my connection string be parameterized too?

I recognize that the key is not the tokenString, I thought the key was the same as a UUID from this snippet but maybe that doesn't have to be the case using EF. **What am I missing, O EF Veteran?**
![image](https://user-images.githubusercontent.com/8525467/126424951-2251a9f2-2c6e-4b9f-9916-245eb7e9cb8b.png)

note to self: the only thing that's missing to test end-to-end is to actually send push notifications by making requests for each device token in the database to Apple's APNS at ~15s before InputRegistration phase ends

Kukks committed:
Refactor EF usage

Kukks commented:
@DanGould  I've opened a PR that refactors the EF usage (https://github.com/DanGould/WalletWasabi/pull/2). Notably I've:
* switched to using the DbContextFactory, which allows you to not have to rely on an HTTP lifecycle to provide a scope for the services
* added a designtime context factory, which lets you generate migrations without needing to hard code the conn string in the context
* added a migration startup task that runs migrations to the db
* renamed the dbcontext to something less specific
* added a new property to Config to load the connection string
* added a docker compose to spin up a postgres db locally

Kukks commented on WalletWasabi.Backend/Controllers/APNTokensController.cs:
```diff
@@ -0,0 +1,54 @@
+		[HttpPost]
+		[ProducesResponseType(200)]
+		[ProducesResponseType(400)]
+		public async Task<IActionResult> StoreTokenAsync([FromBody] AppleDeviceToken token)
```
You need some sort of spam protection here. This is a ddos attack vector.  See https://github.com/NicolasDorier/NicolasDorier.RateLimits for a basic limiter but you should also add some sort of interactive verification because an asshole will spam your db

DanGould replied:
Wasabi suffered [DDoS in June](https://blog.wasabiwallet.io/wve-006-ddos-attack-report/) so I agree that there is a real threat to defend against. At the same time, to coordinate peers in a 30 second window we have to process hundreds / thousands of requests. I'm not sure what the threshold of stability is here and probably won't until we go out try & have things break. By interactive verification, do you mean issuing tokens to authorize these requests?

Kukks replied:
Yes, make every device sending you these requests require some form of auth (eg store a pubkey on backend and have them sign the HTTP body.)

DanGould replied:
I think we've decided to use hashcash instead

DanGould committed:
Merge pull request #2 from Kukks/feat/remote-notifications

Refactor EF usage

DanGould committed:
fix DeleteToken

DanGould committed:
SchedulePush notifications before round input is over

DanGould committed:
create apns push service

Kukks commented on WalletWasabi.Backend/SendPushService.cs:
```diff
@@ -0,0 +1,101 @@
+			//removeToken prepared statement
+			var server = isDebug ? "api.development" : "api";
+			// get options from config? or context from factory?
+			using var context = new WasabiBackendContext(null);
```
```suggestion
			await using var context = _contextFactory.CreateContext();
```

Kukks commented on WalletWasabi.Backend/SendPushService.cs:
```diff
@@ -0,0 +1,101 @@
+				}
+			}";
+
+		public SendPushService()
```
```suggestion
		public SendPushService(IDbContextFactory<WasabiBackendContext> contextFactory)
```

Kukks commented on WalletWasabi.Backend/SendPushService.cs:
```diff
@@ -0,0 +1,101 @@
+	public class SendPushService
+	{
+		private readonly string KeyPath = "/home/Dan/Downloads/AuthKey_4L3728R8LJ.p8";
+		private WasabiBackendContext Context { get; set; }
```
```suggestion
		private readonly IDbContextFactory<WasabiBackendContext> _contextFactory;
```

Kukks commented on WalletWasabi.Backend/SendPushService.cs:
```diff
@@ -0,0 +1,101 @@
+{
+	public class SendPushService
+	{
+		private readonly string KeyPath = "/home/Dan/Downloads/AuthKey_4L3728R8LJ.p8";
```
store in `Config`

Kukks commented on WalletWasabi.Backend/Migrations/APNTokensContextModelSnapshot.cs:
```diff
@@ -0,0 +1,42 @@
+namespace WalletWasabi.Backend.Migrations
+{
+    [DbContext(typeof(WasabiBackendContext))]
+    partial class APNTokensContextModelSnapshot : ModelSnapshot
```
```suggestion
    partial class WasabiBackendContextModelSnapshot : ModelSnapshot
```

Kukks commented on WalletWasabi.Backend/Global.cs:
```diff
@@ -88,7 +88,8 @@ public async Task InitializeAsync(Config config, CoordinatorRoundConfig roundCon
 			var blockNotifier = HostedServices.FirstOrDefault<BlockNotifier>();
 			IndexBuilderService = new IndexBuilderService(RpcClient, blockNotifier, indexFilePath);
-			Coordinator = new Coordinator(RpcClient.Network, blockNotifier, Path.Combine(DataDir, "CcjCoordinator"), RpcClient, roundConfig);
+			var pushService = new SendPushService();
```
pass from DI somehow -- either in Global ctor or in this method param

Kukks commented on WalletWasabi.Backend/Data/DesignTimeDbContextFactoryBase.cs:
```diff
@@ -0,0 +1,71 @@
+
+		private TContext Create(string basePath, string environmentName)
+		{
+			var builder = new ConfigurationBuilder()
```
```suggestion

```

DanGould replied:
????Ｗ吃妮䢎ㄍ

DanGould replied:
oops I mean what is the suggestion here

Kukks commented on WalletWasabi.Backend/Data/DesignTimeDbContextFactoryBase.cs:
```diff
@@ -0,0 +1,71 @@
+			return Create(basePath, environmentName);
+		}
+
+		private TContext Create(string basePath, string environmentName)
```
Replace body of this whole method with `return Create("User ID=postgres;Host=127.0.0.1;Port=65466;Database=doesntmatterbecauseitisnotactuallyused;");`

DanGould replied:
just get rid of basePath and envName?

Kukks commented on WalletWasabi.Backend/SendPushService.cs:
```diff
@@ -0,0 +1,101 @@
+		{
+		}
+
+		private string GenerateAuthenticationHeader()
```
What is this?

DanGould replied:
Apple requires a JWT Authentication header, it's complex unfortunately. I should:

- [ ] leave a comment summary for GenerateAuthenticationHeader

Kukks replied:
Instead of doing all that manually, you can use dotnet libs:

```
var tokenHandler = new JwtSecurityTokenHandler();
            var securityToken = tokenHandler.CreateToken(tokenDescriptor);
            var token = tokenHandler.WriteToken(securityToken);
```

DanGould replied:
I'm not sure it does, we need to sign the request with that private key in the p8 file

DanGould commented:
Andrew what does this mean: https://github.com/DanGould/WalletWasabi/pull/2#discussion_r674143742 @Kukks

DanGould committed:
mv APNTokensContext WasabiBackendContext

DanGould committed:
set startup with magic config

DanGould commented:
- magic strings instead of proper config at this moment cause 
- It's too far gone to save. Redo DI like we did with Chaincase is in order
- we've gone so long without giving users the updates they deserve.
- testnet is going haywire and my mempool will not full, so I'm going to try the end-to-end test tomorrow

Kukks committed:
Refactor some more

Kukks committed:
Parallelize notif push

DanGould committed:
rm duplicate code

DanGould committed:
use literal config since config read after DI

DanGould committed:
read apns p8 file

DanGould committed:
send first successful push from backend
had to rm Task.WhenAll on tokens => SendNotification because it wasn't creating tasks and DbContext is not thread safe

DanGould commented:
@Kukks why are we using postgres at port 65466 instead of the default 5432

Kukks commented:
It is configurable in the config, but I set a non standard port as default
so that you can run the dev docker compose file without any port conflicts
on your dev machine

On Thu, 29 Jul 2021, 21:04 Dan Gould, ***@***.***> wrote:

> @Kukks <https://github.com/Kukks> why are we using postgres at port 65466
> instead of the default 5432
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/DanGould/WalletWasabi/pull/1#issuecomment-889387200>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAN357SWOCE57AQ5DW3AMKTT2GQ3VANCNFSM5AXCBIPQ>
> .
>

DanGould committed:
fix APNs request

DanGould committed:
use || instead of or

DanGould committed:
fix push payload

DanGould committed:
Only schedule 1 notification since 2 rounds are simultaneous

DanGould committed:
spread round timeouts evenly
https://github.com/zkSNACKs/WalletWasabi/pull/5407

DanGould committed:
background notification header

DanGould committed:
pure background notification

DanGould committed:
use reliable push settingsfrom isolated project

DanGould committed:
only SendPush if round.CountAlices() >= 2

DanGould committed:
CountQueuedAlices to control SendPush

DanGould committed:
post ApnsToken

DanGould committed:
Tokens gets Unique constraint

DanGould committed:
DOH (undo edit migrations

DanGould commented:
- [x] Tokens col needs to be unique, else the code that loops to notify notifies the same device id'd by that token multiple times. Perhaps primary key
- [x] Store tokens should return OK if valid token & already in DB rather than 500 on the dupe. catch error

I just set the DB col unique manually to test but we gotta fix the migration for prod

DanGould committed:
push silent & keyed notification

Kukks committed:
change token model

This PR:
* changes the name of the model to be less specific
* removes isDebug and makes an enum Type (allows abstracting the send push service to support additional token formats later if need be, such as android, or perhaps a third party push service if ever possible)
* Calls SendNotificationAsync in parallel instead of waiting for each request to finish
* Makes the token the table's primary key

Note: I generated a new migration but it will not be a smooth update as I did not add logic migrate the data. I would suggest to start fresh.

Kukks committed:
Add Token Status

DanGould committed:
Merge pull request #6 from Kukks/feat/remote-notifications

change token model

DanGould committed:
build rc

DanGould committed:
PUT deviceToken

DanGould committed:
put

DanGould committed:
log APNS request fail content

DanGould committed:
configure notifications isDebug @ coordinator

DanGould committed:
sequentially process notifications
I think the HttpClient gets screwed up using Task.WhenAll(... We're going to have to investigate this further

DanGould committed:
don't send apns-id

DanGould committed:
don't modify the collection you're looping over

DanGould committed:
apns-expiration: 0
tell apns not to retain the notification, only deliver it just exactly once

DanGould commented:
I'd like to be sending notifications with alerts cause then we can set priority=10 (aka DELIVER THIS NOW) to apns. see https://developer.apple.com/documentation/usernotifications/setting_up_a_remote_notification_server/sending_notification_requests_to_apns/. An alert is defined as a **sound**, **badge** or an actual card **alert** that shows up in notification senter

If using a [notification service extension](https://developer.apple.com/documentation/usernotifications/unnotificationserviceextension/1648229-didreceive) with mutable content, you can't change an alert to empty so I'm afraid then EVERYONE in the anonymity set will get an alert and you can't easily drop those. Maybe we can play a silent sound file or something if there's no actual alert to show otherwise show mutable content? or Have a walkthrough in setup to turn sound off?

The best bet may be to send a mutable-content badge with apns-push-type: alert and to then badge the app based on the success of the CoinJoin

DanGould committed:
0 remix registration delay

DanGould committed:
spacing?

Kukks committed:
Anti-spam protection

This PR adds anti spam protection for the new notifications endpoints.  There are actually 2 mechanisms introduced here:
* A hashcash implementation where the client must now provide a X-Hashcash header with a v1 hashcash stamp with sufficient pow
* A rate limiter where: token submission is throttled based on user ip and token deletion is throttled by the token being deleted

Kukks committed:
Add Notification methods to Client and tests

Kukks committed:
Add methods of token to client

Kukks committed:
fix: hash cash utils verify

Kukks committed:
Prevent double spends of hashcash computes

Kukks committed:
Make sure to use 3.1 libs

Tests would not work with the 5.0  libs

Kukks committed:
pow of 20 is the default from spec

This is really not ideal for prod and should be adjusted in production based on demand. We can potentially just add a config key and make it reload on change from file instead of advanced ddos detection mechanism

Kukks committed:
Read Pow for hashcash from config

Kukks committed:
Adapt to use Http Hashcash

Kukks commented on WalletWasabi.Backend/Controllers/ChaumianCoinJoinController.cs:
```diff
@@ -302,17 +302,19 @@ public async Task<IActionResult> PostInputsAsync([FromBody, Required]InputsReque
 						acceptedBlindedOutputScripts.Add(blindedOutputs[i]);
 					}
 
+
```
```suggestion
```

Kukks commented on WalletWasabi.Backend/Controllers/ChaumianCoinJoinController.cs:
```diff
@@ -302,17 +302,19 @@ public async Task<IActionResult> PostInputsAsync([FromBody, Required]InputsReque
+					// if alice is counted in queued count already, remove it
+					round.DequeueAnyFamiliarAlice(request.Inputs);
 					round.AddAlice(alice);
+
```
```suggestion
```

Kukks commented on WalletWasabi.Backend/Global.cs:
```diff
@@ -88,11 +89,12 @@ public async Task InitializeAsync(Config config, CoordinatorRoundConfig roundCon
 			IndexBuilderService = new IndexBuilderService(RpcClient, blockNotifier, indexFilePath);
-			Coordinator = new Coordinator(RpcClient.Network, blockNotifier, Path.Combine(DataDir, "CcjCoordinator"), RpcClient, roundConfig);
+			Coordinator = new Coordinator(RpcClient.Network, blockNotifier, Path.Combine(DataDir, "CcjCoordinator"), RpcClient, roundConfig,
+				() => _ = serviceProvider.GetRequiredService<SendPushService>().SendNotificationsAsync(roundConfig.IsDebug));
```
this func for lazy DI shows that there is a design flaw somewhere. We should fix this later on

DanGould replied:
yep, wasabi.backend doesn't use DI.

Kukks commented on WalletWasabi.Backend/Migrations/20210825090127_ChangeTokenModel.cs:
```diff
@@ -0,0 +1,71 @@
+
+namespace WalletWasabi.Backend.Migrations
+{
+    public partial class ChangeTokenModel : Migration
```
No need for a migration from the start: delete and db seed files and create the first migration again

Kukks commented on WalletWasabi.Backend/WalletWasabi.Backend.csproj:
```diff
@@ -38,11 +38,28 @@
       <GlobalPropertiesToRemove></GlobalPropertiesToRemove>
     </ProjectReference>
   </ItemGroup>
+  <ItemGroup>
```
junk?

DanGould replied:
yes

Kukks commented on WalletWasabi.Standard/CoinJoin/Common/Models/RoundStateResponseBase.cs:
```diff
@@ -22,10 +22,10 @@ public abstract class RoundStateResponseBase
 
 		public int RegisteredPeerCount { get; set; }
 
-		public int QueuedPeerCount { get; set; }
```
revert to reduce diff with upstream

Kukks commented on WalletWasabi.Standard/CoinJoin/Coordinator/Coordinator.cs:
```diff
@@ -216,38 +218,50 @@ public async Task ProcessConfirmedTransactionAsync(Transaction tx)
+					round.PhaseChanged += Round_PhaseChangedAsync;
+					await round.ExecuteNextPhaseAsync(RoundPhase.InputRegistration).ConfigureAwait(false);
 					Rounds.Add(round);
+					SchedulePushAsync(round.InputRegistrationTimeout);
```
Instead of the `#pragma warning disable` you can do 
```suggestion
					_ = SchedulePushAsync(round.InputRegistrationTimeout);
```

Kukks commented on WalletWasabi.Standard/CoinJoin/Coordinator/Coordinator.cs:
```diff
@@ -216,38 +218,50 @@ public async Task ProcessConfirmedTransactionAsync(Transaction tx)
+#pragma warning restore CS4014 // Because this call is not awaited, execution of the current method continues before the call is completed
+
+		public async Task SchedulePushAsync(TimeSpan inputRegistrationTimeout)
+		{
```
Does it need to be this close to the timeout?

Kukks replied:
You are deliberately waiting to be within 15s of the deadline to register inputs?

DanGould replied:
Yes, this is synchronous sending of a push notification. It takes ~15s from the time the notification is sent til client input registration success, a prerequisite for the connection confirmatino, output registration, & signing phases. Even if a client registers for a CoinJoin, when they lose connection they lose registration and so must register again here.

Reducing the time to register is one of our most probable opportunities to stabilize background coinjoin in my estimation.

Kukks commented on WalletWasabi.Standard/CoinJoin/Coordinator/Coordinator.cs:
```diff
@@ -216,38 +218,50 @@ public async Task ProcessConfirmedTransactionAsync(Transaction tx)
+			{
+				// this sends 2 notifications for two rounds, so make sure the notifications are rate limited
+				Logger.LogInfo("PUSH NOTIFICATIONS! INPUT DONE IN 15.");
+				SendPushCallback.Invoke();
```
I would do an event instead of an Action callback

Kukks commented on WalletWasabi.Backend/Startup.cs:
```diff
@@ -62,9 +65,23 @@ public void ConfigureServices(IServiceCollection services)
+
+			services.AddDbContextFactory<WasabiBackendContext>((provider, builder) =>
+			{
+				var connString = "User ID=postgres;Host=127.0.0.1;Port=65466;Database=wasabibackend;";
```
```suggestion
				var connString = provider.GetService<Global>().Config.DatabaseConnectionStringName;
```

Kukks committed:
fix db factory polyfills

Kukks commented on WalletWasabi.Backend/WalletWasabi.Backend.csproj:
```diff
@@ -38,11 +38,28 @@
   <ItemGroup>
     <PackageReference Include="Swashbuckle.AspNetCore" Version="5.0.0" />
     <PackageReference Include="Microsoft.AspNetCore.Mvc.NewtonsoftJson" Version="3.1.6" />
+    <PackageReference Include="Npgsql.EntityFrameworkCore.PostgreSQL" Version="5.0.7" />
```
5.0 packages in 3.1 is gonna cause some headaches. I fixed this in https://github.com/DanGould/WalletWasabi/pull/7

DanGould replied:
- [x] merge #7

DanGould committed:
Update WalletWasabi.Backend/Controllers/ChaumianCoinJoinController.cs

Co-authored-by: Andrew Camilleri <evilkukka@gmail.com>

DanGould committed:
Update WalletWasabi.Backend/Controllers/ChaumianCoinJoinController.cs

Co-authored-by: Andrew Camilleri <evilkukka@gmail.com>

Kukks committed:
change name of diff var

DanGould committed:
Merge pull request #7 from Kukks/feat-anti-spam

Anti-spam protection -- delete endpoint is superfluous

DanGould committed:
Merge branch 'feat/remote-notifications' of https://github.com/dangould/WalletWasabi into feat/remote-notifications

DanGould committed:
Remove junk in csproj

DanGould committed:
Remove junk in csproj

DanGould commented on WalletWasabi.Backend/docker-compose.yml:
```diff
@@ -0,0 +1,14 @@
+  postgres:
+    image: postgres:13.1
+    environment:
+      POSTGRES_HOST_AUTH_METHOD: trust
```
🚩?

Kukks replied:
This Docker compose is for DEV ENVIRONMENT ONLY

DanGould replied:
where's our setup for prod?

DanGould committed:
Remove junk in csproj

Kukks committed:
Replace with single migration

DanGould committed:
Merge pull request #8 from Kukks/single-migration

Replace with single migration

DanGould commented:
I realize now this is still missing the v4 hard fork backend upgrade

https://github.com/zkSNACKs/WalletWasabi/pull/3739
